// @wdio/local-runner/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _logger = _interopRequireDefault(require("@wdio/logger"));

var _streamBuffers = require("stream-buffers");

var _worker = _interopRequireDefault(require("./worker"));

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const log = (0, _logger.default)('@wdio/local-runner');

class LocalRunner {
  constructor(configFile, config) {
    this.configFile = configFile;
    this.config = config;
    this.workerPool = {};
    this.stdout = new _streamBuffers.WritableStreamBuffer(_constants.BUFFER_OPTIONS);
    this.stderr = new _streamBuffers.WritableStreamBuffer(_constants.BUFFER_OPTIONS);
  }

  initialise() {}

  getWorkerCount() {
    return Object.keys(this.workerPool).length;
  }

  run({
    command,
    argv,
    ...options
  }) {
    const workerCnt = this.getWorkerCount();

    if (workerCnt >= process.stdout.getMaxListeners() - 2) {
      process.stdout.setMaxListeners(workerCnt + 2);
      process.stderr.setMaxListeners(workerCnt + 2);
    }

    const worker = new _worker.default(this.config, options, this.stdout, this.stderr);
    this.workerPool[options.cid] = worker;
    worker.postMessage(command, argv);
    return worker;
  }

  shutdown() {
    log.info('Shutting down spawned worker');

    for (const [cid, worker] of Object.entries(this.workerPool)) {
      const {
        caps,
        server,
        sessionId,
        config,
        isMultiremote,
        instances
      } = worker;
      let payload = {};

      if (config && config.watch && (sessionId || isMultiremote)) {
        payload = {
          config: { ...server,
            sessionId
          },
          caps,
          watch: true,
          isMultiremote,
          instances
        };
      } else if (!worker.isBusy) {
        delete this.workerPool[cid];
        continue;
      }

      worker.postMessage('endSession', payload);
    }

    return new Promise(resolve => {
      const interval = setInterval(() => {
        const busyWorker = Object.entries(this.workerPool).filter(([, worker]) => worker.isBusy).length;
        log.info(`Waiting for ${busyWorker} to shut down gracefully`);

        if (busyWorker === 0) {
          clearInterval(interval);
          log.info('shutting down');
          return resolve();
        }
      }, 250);
      setTimeout(resolve, _constants.SHUTDOWN_TIMEOUT);
    });
  }

}

exports.default = LocalRunner;
